### ISSUE TITLE
Helm config for ACL token distribution to K8s nodes

### ISSUE THREAD

**Issue by plotia-charter** [hashicorp/consul#7578]
I'm building a hybrid environment with a bare-metal Consul server node and clients who are a part of a K8s cluster. My goal here is to get Mesh Gateway setup. I have reviewed the guides on https://learn.hashicorp.com
I have a single node as consul server (no redundancy setup yet), 3 K8s nodes as clients (1 master & 2 slaves) and a another bare-metal consul server for a second DC. I have WAN joined the two consul server nodes.

While configuring Mesh Gateway, I first need to configure ACL and then ACL replication. I'm working on setting up ACLs right now.

After reading the documentation and doing the configuration, I have still not understood a few concepts. 

1. Does a single helm chart spin-up consul clients on all K8s nodes? My understanding is that it does as the consul members list does show up the K8s nodes.
2. How are ACL node tokens assigned to various K8s nodes via a single helm chart?

Here is my consul server config (it uses 1.7.2):
```
{
  "datacenter": "dc1",
  "primary_datacenter": "dc1",
  "data_dir": "/tmp/consul",
  "log_level": "INFO",
  "node_name": "server1",
  "server": true,
  "connect": {
    "enabled": true,
  },
  "ports": {
    "https": 8501,
    "grpc": 8502
  },
  "bind_addr": "10.0.0.1",
  "client_addr": "10.0.0.1"
  "bootstrap_expect": 1,
  "ui": true,
  "enable_local_script_checks": true
  "acl": {
    "enabled": true,
    "down_policy": "async-cache",
    "default_policy": "deny",
    "enable_token_persistence": true
  }
}
```

Here is the helm config on the K8s master node:
```
global:
  enabled: false

client:
  enabled: true
  image: "consul:1.7.2"
  exposeGossipPorts: true
  join:
    - "10.0.0.1"
  grpc: true

connectInject:
  enabled: true
  default: true
  imageEnvoy: "envoyproxy/envoy-alpine:v1.13.0"
  aclInjectToken:
    secretName: null       //reserved this for service token for ACL
    secretKey: null          //reserved this for service token for ACL
  aclBindingRuleSelector: ""
  centralConfig:
    enabled: true
    defaultProtocol: "tcp"

syncCatalog:
  enabled: true
  default: true
  aclSyncToken:
    secretName: "secret-k8s-master-agent"     //this is where the node ACL token is stored
    secretKey: "secret_k8s_master_agent".        //this is where the node ACL token is stored
  logLevel: "trace"

tests:
  enabled: true
```

Here is my secret definition file which defines token for the K8s master node:
```
apiVersion: v1
kind: Secret
metadata:
  name: secret-k8s-master-agent 
type: Opaque
data:
  secret_k8s_master_agent: <redacted-base64-of-token>
```

How and where do I add ACL node tokens for other K8s nodes which I want as clients?

**Comment by jsosulska** [comment]
Hi @plotia-charter 

Thanks for this really detailed question, this does help add a lot of context! 

> Does a single helm chart spin-up consul clients on all K8s nodes? My understanding is that it does as the consul members list does show up the K8s nodes.

In Consul's Helm installation guide, we point out that the [Client Agent](https://www.consul.io/docs/platform/k8s/run.html#client-agents) have a few different configurations. I've included the blurb below, but there is more information about client agents also provided.

>The client agents are run as a DaemonSet. This places one agent (within its own pod) on each Kubernetes node. The clients expose the Consul HTTP API via a static port (default 8500) bound to the host port. This enables all other pods on the node to connect to the node-local agent using the host IP that can be retrieved via the Kubernetes downward API.  

> How are ACL node tokens assigned to various K8s nodes via a single helm chart?
> How and where do I add ACL node tokens for other K8s nodes which I want as clients?

I don't know enough Consul-helm to be able to answer this, so I will do some more digging, or maybe someone from the community can help. What is your use case? Have you tried setting up the [Kubernetes Auth Method](https://www.consul.io/docs/acl/auth-methods/kubernetes.html)? 

Also, some folks may not see this particular question. If you are looking to ask more questions, I recommend [our Discuss forums for Consul](https://discuss.hashicorp.com/c/consul/29), where many community members can chime in :) 

I hope this helps a little bit. I'll see if I can find more information.
Best,
Jono

**Comment by plotia-charter** [comment]
@jsosulska Thank you for looking into this and sending me details.
I haven't setup K8s Auth method but I was able to get in touch with Consul support regarding the token related questions. I will post the answer here so folks can find it in future.

The short answer is that all policies for a cluster should be created using a single policy file. Then a single token can be generated for that and that is what is used in the helm chart. This is opposite to what I thought - declaring a single policy per node per file and then generating multiple tokens.
This is how you add the ACL token in the helm-chart. Generate the agent policy on the consul server and generate a token for it. The token will be added below. Under the client config part in the helm chart, add the following:

```
 extraConfig: |
    {
      "acl": {
        "enabled": true,
        "default_policy": "deny",
        "down_policy": "extend-cache",
          "tokens": {
            "agent": "<generated-token-here>"
        }
      }
    }
```

Additionally, create a policy for the 'k8s-sync' node which is a built-in node for K8s sync operations. Example below:

```
node "k8s-sync" {
    policy = "write"
  }
node_prefix "" {
    policy = "read"
  }
service_prefix "" {
    policy = "write"
  }
```

Generate a token for this policy and use the token as a K8s secret to reference it in the helm-chart for 'aclSyncToken' values.

I think I've got my question answered and I will be closing this issue.
